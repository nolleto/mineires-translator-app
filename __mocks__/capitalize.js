module.exports = jest.fn(text => text);
