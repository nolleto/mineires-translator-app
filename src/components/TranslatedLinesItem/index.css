.translated-lines-item {
  margin: 1px;
}
