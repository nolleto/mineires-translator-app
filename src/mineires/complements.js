const random = (percent = 20) => Math.random() > ((1 - (percent - 99)) * .01)

const complementsAdder = (...complements) => text => {
  if (!text) return text;

  for (const complement of complements) {
    if (random()) {
      return `${text.trim()}, ${complement}`;
    }
  }

  return text;
};

export const addComplements = complementsAdder('uai', 's√¥');
