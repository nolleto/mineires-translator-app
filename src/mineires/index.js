import {
  creatorPhraseReplaces,
  creatorSuffixReplaces,
  creatorWordReplaces,
  executeReplacesInText
} from './replacer';

import PHRASES from './constants/phrases.json';
import SUFFIXS from './constants/suffixs.json';
import WORDS from './constants/words.json';

const phraseReplaces = creatorPhraseReplaces(PHRASES);
const wordReplaces = creatorWordReplaces(WORDS);
const suffixWords = creatorSuffixReplaces(SUFFIXS);

export default executeReplacesInText(
  phraseReplaces,
  wordReplaces,
  suffixWords,
);
