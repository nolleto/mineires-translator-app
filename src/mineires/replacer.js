import ACCENTS from './constants/accents.json';

const BEGIN_PHRASE = '(^| )'
const END_PHRASE = '( |\\.|,|!|\\?|$)'

const stringToRegex = str => new RegExp(str, 'ig');
const createPhraseRegex = phrase => stringToRegex(replaceAccentsInString(`${BEGIN_PHRASE}(${phrase})${END_PHRASE}`));
const createWordRegex = word => stringToRegex(replaceAccentsInString(`${BEGIN_PHRASE}(${word})${END_PHRASE}`));
const createSuffixRegex = word => stringToRegex(`([a-z]+)(${replaceAccentsInString(word)})${END_PHRASE}`);

const replaceAccentsInString = str => {
  return Object.keys(ACCENTS).reduce((acc, letter) => {
    return acc.replace(new RegExp(letter, 'ig'), ACCENTS[letter]);
  }, str);
};

const getTargetPhrases = phrases => {
  if (Array.isArray(phrases)) return phrases;
  return [phrases];
};

export const creatorPhraseReplaces = json => text => {
  return Object.keys(json).reduce((textToReplace, newPhrase) => {
    const targetPhrases = getTargetPhrases(json[newPhrase]);

    return targetPhrases.reduce((acc, targetPhrase) => {
      const regex = createPhraseRegex(targetPhrase);

      return acc.replace(regex, `$1${newPhrase}$3`);
    }, textToReplace);
  }, text);
};

export const creatorWordReplaces = json => text => {
  return Object.keys(json).reduce((textToReplace, newWord) => {
    const targetWords = getTargetPhrases(json[newWord]);

    return targetWords.reduce((acc, targetWord) => {
      const regex = createWordRegex(targetWord);

      return acc.split(' ').map(
        word => word.replace(regex, `$1${newWord}$3`)
      ).join(' ');
    }, textToReplace);
  }, text);
};

export const creatorSuffixReplaces = json => text => {
  return Object.keys(json).reduce((textToReplace, newWord) => {
    const targetWords = getTargetPhrases(json[newWord]);

    return targetWords.reduce((acc, targetWord) => {
      const regex = createSuffixRegex(targetWord);

      return acc.replace(regex, `$1${newWord}$3`);
    }, textToReplace);
  }, text);
};

export const executeReplacesInText = (...replaces) => text => {
  return replaces.reduce((acc, replace) => replace(acc), text);
};
