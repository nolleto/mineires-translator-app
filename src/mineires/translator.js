import {
  creatorPhraseReplaces,
  creatorSuffixReplaces,
  creatorWordReplaces,
  executeReplacesInText
} from './replacer';

import PHRASES from './constants/phrases.json';
import SUFFIXS from './constants/suffixs.json';
import WORDS from './constants/words.json';

export default executeReplacesInText(
  creatorPhraseReplaces(PHRASES),
  creatorWordReplaces(WORDS),
  creatorSuffixReplaces(SUFFIXS),
);
