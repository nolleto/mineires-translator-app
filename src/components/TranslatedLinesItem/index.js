import './index.css';

import PropTypes from 'prop-types';
import React from 'react';
import capitalize from 'capitalize';
import toMineires from '../../mineires';

const TranslatedLinesItem = ({ text }) => (
  <p className='translated-lines-item'>
    {capitalize(toMineires(text))}
  </p>
);

TranslatedLinesItem.propTypes = {
  text: PropTypes.string.isRequired
};

export default TranslatedLinesItem;
