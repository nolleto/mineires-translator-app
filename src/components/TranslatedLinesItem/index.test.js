import * as capitalize from 'capitalize';
import * as translator from '../../mineires/translator';

import { mount, shallow } from 'enzyme';

import React from 'react';
import TranslatedLinesItem from './index';

const shallowComponent = (props = {}) => {
  const finalProps = {
    text: '',
    ...props
  };

  return shallow(
    <TranslatedLinesItem
      {...finalProps}
    />
  );
};

const mountComponent = (props = {}) => {
  const finalProps = {
    text: '',
    ...props
  };

  return mount(
    <TranslatedLinesItem
      {...finalProps}
    />
  );
};

describe('TranslatedLinesItem component', () => {
  describe('Snapshot', () => {
    it('when has no text', () => {
      const snapshot = shallowComponent({ text: '' });

      expect(snapshot).toMatchSnapshot();
    });

    it('when has text', () => {
      const snapshot = shallowComponent({
        text: 'some text'
      });

      expect(snapshot).toMatchSnapshot();
    });
  });

  describe('Text parser', () => {
    beforeAll(() => {
      translator.default = jest.fn(text => text);
      mountComponent({ text: 'some text' });
    });

    it('calls capitalize', () => {
      expect(capitalize.default).toBeCalled();
    });

    it('calls mineires translator', () => {
      expect(translator.default).toBeCalled();
    });
  });
});
