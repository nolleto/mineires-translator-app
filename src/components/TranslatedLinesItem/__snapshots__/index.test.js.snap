// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TranslatedLinesItem component Snapshot when has no text 1`] = `
<p
  className="translated-lines-item"
/>
`;

exports[`TranslatedLinesItem component Snapshot when has text 1`] = `
<p
  className="translated-lines-item"
>
  some text
</p>
`;
